google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,239 @@
+			scriptsReturnCode, agentErrorCode = status.ExitStatus(), pb.AgentErrorCode_UNHANDLED_SCRIPT_ERROR
+		}
+	}
+
+	if config.ContinueOnScriptError {
+		agentErrorCode = pb.AgentErrorCode_NO_ERROR
+	}
+
+	return scriptsReturnCode, agentErrorCode
+}
+
+func listenForSnapshotRequests(address string, requestChan chan<- *pb.GuestMessage) {
+	for {
+		// Start hanging connection on server that feeds to channel
+		logger.Infof("attempting to connect to snapshot service at %s", address)
+		conn, err := grpc.Dial(address, grpc.WithInsecure())
+		if err != nil {
+			logger.Errorf("failed to connect: %v", err)
+			continue
+		}
+
+		c := pb.NewSnapshotServiceClient(conn)
+		ctx, cancel := context.WithCancel(context.Background())
+		guestReady := pb.GuestReady{
+			RequestServerInfo: false,
+		}
+		r, err := c.CreateConnection(ctx, &guestReady)
+		if err != nil {
+			logger.Errorf("error creating connection: %v", err)
+			continue
+		}
+		for {
+			request, err := r.Recv()
+			if err != nil {
+				logger.Errorf("error reading snapshot request: %v", err)
+				break
+			}
+			logger.Infof("received snapshot request")
+			requestChan <- request
+		}
+		cancel()
+	}
+}
+
+func getSnapshotResponse(guestMessage *pb.GuestMessage) *pb.SnapshotResponse {
+	switch {
+	case guestMessage.GetSnapshotRequest() != nil:
+		request := guestMessage.GetSnapshotRequest()
+		response := &pb.SnapshotResponse{
+			OperationId: request.GetOperationId(),
+			Type:        request.GetType(),
+		}
+
+		config, err := getSnapshotConfig()
+		if err != nil {
+			response.AgentReturnCode = pb.AgentErrorCode_INVALID_CONFIG
+			return response
+
+		}
+
+		var url string
+		switch request.GetType() {
+		case pb.OperationType_PRE_SNAPSHOT:
+			logger.Infof("handling pre snapshot request for operation id %d", request.GetOperationId())
+			if seenPreSnapshotOperationIds.Contains(request.GetOperationId()) {
+				logger.Infof("duplicate pre snapshot request operation id %d", request.GetOperationId())
+				return nil
+			}
+			seenPreSnapshotOperationIds.Add(request.GetOperationId(), nil)
+			url = config.PreSnapshotScriptURL
+		case pb.OperationType_POST_SNAPSHOT:
+			logger.Infof("handling post snapshot request for operation id %d", request.GetOperationId())
+			if seenPostSnapshotOperationIds.Contains(request.GetOperationId()) {
+				logger.Infof("duplicate post snapshot request operation id %d", request.GetOperationId())
+				return nil
+			}
+			seenPostSnapshotOperationIds.Add(request.GetOperationId(), nil)
+			url = config.PostSnapshotScriptURL
+		default:
+			logger.Errorf("unhandled operation type %d", request.GetType())
+			return nil
+		}
+
+		scriptsReturnCode, agentErrorCode := runScript(url, request.GetDiskList(), config)
+		response.ScriptsReturnCode = int32(scriptsReturnCode)
+		response.AgentReturnCode = agentErrorCode
+
+		return response
+	default:
+	}
+	return nil
+}
+
+func handleSnapshotRequests(address string, requestChan <-chan *pb.GuestMessage) {
+	for {
```

Comment:
same qn here - why are you infinitely creating new connections? can't handler dial up connections only when there are messages on the channel?

Reply:
Same thing, was worried about reconnecting.